Sources/ArgumentParser/Completions/BashCompletionsGenerator.swift

```diff
@@ -9,257 +9,397 @@
+    // - first is guaranteed non-empty in the one place where this computed var is used.
+    let commandName = first!._commandName
     return """
       #!/bin/bash
 
-      \(generateCompletionFunction([type]))
+      # positional arguments:
+      #
+      # - 1: the current (sub)command's count of positional arguments
+      #
+      # required variables:
+      #
+      # - flags: the flags that the current (sub)command can accept
+      # - options: the options that the current (sub)command can accept
+      # - positional_number: value ignored
+      # - unparsed_words: unparsed words from the current command line
+      #
+      # modified variables:
+      #
+      # - flags: remove flags for this (sub)command that are already on the command line
+      # - options: remove options for this (sub)command that are already on the command line
+      # - positional_number: set to the current positional number
+      # - unparsed_words: remove all flags, options, and option values for this (sub)command
+      \(offerFlagsOptionsFunctionName)() {
+          local -ir positional_count="${1}"
+          positional_number=0
+
+          local was_flag_option_terminator_seen=false
+          local is_parsing_option_value=false
+
+          local -ar unparsed_word_indices=("${!unparsed_words[@]}")
+          local -i word_index
+          for word_index in "${unparsed_word_indices[@]}"; do
+              if "${is_parsing_option_value}"; then
+                  # This word is an option value:
+                  # Reset marker for next word iff not currently the last word
+                  [[ "${word_index}" -ne "${unparsed_word_indices[${#unparsed_word_indices[@]} - 1]}" ]] && is_parsing_option_value=false
+                  unset "unparsed_words[${word_index}]"
+                  # Do not process this word as a flag or an option
+                  continue
+              fi
+
+              local word="${unparsed_words["${word_index}"]}"
+              if ! "${was_flag_option_terminator_seen}"; then
+                  case "${word}" in
+                  --)
+                      unset "unparsed_words[${word_index}]"
+                      # by itself -- is a flag/option terminator, but if it is the last word, it is the start of a completion
+                      if [[ "${word_index}" -ne "${unparsed_word_indices[${#unparsed_word_indices[@]} - 1]}" ]]; then
+                          was_flag_option_terminator_seen=true
+                      fi
+                      continue
+                      ;;
+                  -*)
+                      # ${word} is a flag or an option
+                      # If ${word} is an option, mark that the next word to be parsed is an option value
+                      # TODO: handle joined-value options (-o=file.ext), stacked flags (-aBc), legacy long (-long), combos
+                      # TODO: if multi-valued options can exist, support them
```

Comment:
Let's move these implementation notes out of the generated script – if there's a place for them in the Swift source, that's okay; otherwise let's just capture in issues.

Reply:
Makes sense. I put the notes in situ since I used them as a guide for what to work on. Once the code is in GitHub, however, I can use code links in issues, which I couldn't do before.